integration/main.go

```diff
@@ -0,0 +1,208 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"os"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/uuid"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/pkg/api/v1"
+	"k8s.io/client-go/tools/clientcmd"
+
+	"github.com/golang/glog"
+	"github.com/spf13/cobra"
+)
+
+const integrationTestNamespacePrefix = "istio-ca-integration-"
+
+var (
+	rootCmd = &cobra.Command{
+		Run:     runTests,
+		PreRun:  initializeIntegrationTest,
+		PostRun: cleanUpIntegrationTest,
+	}
+
+	opts options
+)
+
+type options struct {
+	containerHub   string
+	containerImage string
+	containerTag   string
+	clientset      kubernetes.Interface
+	kubeconfig     string
+	namespace      string
+}
+
+func init() {
+	flags := rootCmd.Flags()
+	flags.StringVar(&opts.containerHub, "hub", "", "Docker hub that the Istio CA image is hosted")
+	flags.StringVar(&opts.containerImage, "image", "", "Name of Istio CA image")
+	flags.StringVar(&opts.containerTag, "tag", "", "Tag for Istio CA image")
+	flags.StringVarP(&opts.kubeconfig, "kube-config", "k", "~/.kube/config", "path to kubeconfig file")
+}
+
+func main() {
+	if err := rootCmd.Execute(); err != nil {
+		glog.Error(err)
+		os.Exit(-1)
+	}
+}
+
+func initializeIntegrationTest(cmd *cobra.Command, args []string) {
+	opts.clientset = createClientset(opts.kubeconfig)
+	opts.namespace = createTestNamespace(opts.clientset)
+
+	deployIstioCA(opts.clientset)
+}
+
+func runTests(cmd *cobra.Command, args []string) {
+	// Test the existence of istio.default.secret.
+	_, err := waitForSecretExist("istio.default", 10*time.Second)
+	if err != nil {
+		glog.Fatal(err)
+	} else {
+		glog.Infof(`Secret "istio.default" is correctly created`)
+	}
+}
+
+func cleanUpIntegrationTest(cmd *cobra.Command, args []string) {
+	deleteTestNamespace(opts.clientset)
+}
+
+func createClientset(kubeconfig string) *kubernetes.Clientset {
+	config, err := clientcmd.BuildConfigFromFlags("", kubeconfig)
+	if err != nil {
+		glog.Errorf("failed to create config object from kube-config file: %q (error: %v)", kubeconfig, err)
```

Comment:
A return here otherwise if we get an error and config is nil, the next lines are going to dereference a nil config.